## 34장 과제 
### 알고리즘을 선택하여 데이타의 class 분류 모델
#### - 보고서 포함 사항
- 데이터 요약
- 데이터 전처리
- 알고리즘별 실험과정 및 결과 설명
- 총 정리

In [7]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from random import *
import sklearn
import sklearn.neural_network as nn
import sklearn.svm as svm
import sklearn.linear_model as lm
import sklearn.neighbors as neighbors
import matplotlib.pyplot as plt

### 1. Data 요약

In [8]:
rawdata =pd.read_csv('sklearn_practice.csv')
rawdata.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [6]:
rawdata.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  3.919560e-15  5.688174e-16 -8.769071e-15  2.782312e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean  -1.552563e-15  2.010663e-15 -1.694249e-15 -1.927028e-16 -3.137024e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.537294e-16  7.959909e-16  5.367590e-16  4.458112e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   1.453003e-15  1.699104e-15 -3.660161e-16 -1.206049e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [7]:
rawdata.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

### 2. 전처리과정
- class_0: 정상처리, class_1: 신용카드사고발생/ 각각 492개만 뽑아서 약 1000개의 데이타로 작업할 것임
- 데이타는 train(약 700개)/test(약 300개)로 나누어서 train으로 학습하여 프로그램을 만들도록 하고 이 후에 test값을 대입하여 신뢰성확인

In [21]:
# 현재 data는 class 0 data 비율이 99.8%이상이므로 결과에 왜곡 발생 가능. class0과 class1의 데이타를 1:1 비율로 구성해서 data 활용필요
class_0 = rawdata[rawdata['Class'] ==0]
class_1  = rawdata[rawdata['Class'] ==1]

#test 용 data에 class 0과 class 1의 비율이 1:1로 섞일 수 있도록 전처리
data = pd.concat([class_0, class_1])
test_data = pd.concat([data.iloc[:100], data.iloc[-100:]])
test_data = test_data.sample(frac=1)
x_test = test_data.loc[:,'Time':'Amount']
y_test = test_data.loc[:,'Class']


#train용 data 전처리
train_data = data.iloc[100:-100]
train_data = train_data.sample(frac=1)
x_train = train_data.loc[:,'Time':'Amount']
y_train = train_data.loc[:,'Class']


### 3. 알고리즘별 실험과정 및 결과 설명
+ 알고리즘 1: LogisticRegression

In [22]:
import sklearn.linear_model as lm

model = lm.LogisticRegression()

model.fit(x_train, y_train) #train data들로 알고리즘1을 학습시킨다.
y_pred = model.predict(x_test) #학습결과(또 다른 프로그램)에 test할 x값을 넣어 예측값 y를 구한다.

cnt=0
for i in range(len(y_pred)): #예측값 y와 실제 y값의 동일여부를 판단하여, 알고리즘1의 예측도를 판별한다
    if y_pred[i] == y_test.iloc[i]:
        cnt+=1
print(cnt/len(y_pred))

C:\Users\queji\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.605


+ 알고리즘 2: SVM

In [23]:
#time을 빼고 V1부터 데이타를 취득한다 (컬럼선택에 따라 결과 신뢰성이 많이 달라짐)
x_train = train_data.loc[:,'V1':'Amount']
y_train = train_data.loc[:,'Class']

x_test = test_data.loc[:,'V1':'Amount']
y_test = test_data.loc[:,'Class']

In [24]:
import sklearn.svm as svm

model = svm.SVC(gamma='scale')

model.fit(x_train, y_train)
y_pred = model.predict(x_test)


cnt=0
for i in range(len(y_pred)):
    if y_pred[i] == y_test.iloc[i]:
        cnt+=1
print(cnt/len(y_pred))

0.53


+ 알고리즘 3: KNN

In [25]:
x_train = train_data.loc[:,'Time':'Amount']
y_train = train_data.loc[:,'Class']

x_test = test_data.loc[:,'Time':'Amount']
y_test = test_data.loc[:,'Class']

In [26]:
import sklearn.neighbors as neighbors

model = neighbors.KNeighborsClassifier()

model.fit(x_train, y_train)
y_pred = model.predict(x_test)


cnt=0
for i in range(len(y_pred)):
    if y_pred[i] == y_test.iloc[i]:
        cnt+=1
print(cnt/len(y_pred))

0.5


+ 알고리즘 4: DecisionTree

In [27]:
import sklearn.tree as tree

model = tree.DecisionTreeClassifier()

model.fit(x_train, y_train)
y_pred = model.predict(x_test)

cnt=0
for i in range(len(y_pred)):
    if y_pred[i] == y_test.iloc[i]:
        cnt+=1
print(cnt/len(y_pred))

0.795


+ 알고리즘 5: MLPClassifier()

In [28]:
import sklearn.neural_network as nn


model = nn.MLPClassifier(learning_rate_init=0.01)
model.fit(x_train, y_train)
y_pred = model.predict(x_test)

cnt=0
for i in range(len(y_pred)):
    if y_pred[i] == y_test.iloc[i]:
        cnt+=1
print(cnt/len(y_pred))

0.5


## 결론.
### 각 machine learning algorithm에 따른 예측 정확도 (본 과제 data에 한정하여)
- 알고리즘 1: LogisticRegression 0.605
- 알고리즘 2: SVM 0.53
- 알고리즘 3: KNN 0.5
- 알고리즘 4: DecisionTree 0.795
- 알고리즘 5: MLPClassifier() 0.5
#### DecisionTree (0.795) > LogisticRegression (0.605) > SVM (0.53) > KNN (0.5) = MLPClassifier() (0.5) 순서로 예측정확도가 높음